### PR TITLE
Mnist - Notebook

### PR DESCRIPTION
**Pull request by TuSKan**
I just scratch a notebook based on Cifar

Changes:
- Create a notebook bases on cifar nb
- removed flagModel and flagLoss from demo.go and from TrainModel. Getting from context.
- Added Depth = 3 constant
- classifier: just adapt the classifier from cifar/classifier
- Remove Loss map function. Creating lossFn from losses.LoadFromContext
- Implement ReduceMinVJP grad func
- Complete losses.LoadFromContext with Triplet and other loss functions
- Implement Softplus activation for better stability of triplet
- Fix  _GetParamOr_,  when hiperparameter is a string: a _v.CanConvert(typeOfT)_  was build only for float32/float64. Using UnmarshalText its possible to convert string "hard" to TripletLossMiningStrategyHard type.

Erros:
- classifier.New is not working. _ Must not error: requested variable "weights" in scope "/model/model/000_conv/conv" with Context.Reuse set, but variable does not exist_.  I suppose to be the weights os the network, but it is not on checkpoints

### CODE REVIEW

File: examples/mnist/classifier/classifier.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
+// Package classifier is a Cifar-10 classifier.
```
Comment (janpfeifer) [inline examples/mnist/classifier/classifier.go]: "...  is an MNIST-based digit classifier."

or something like that?

File: graph/ops_util.go [hunk 2]
```diff
@@ -409,6 +425,29 @@ func MaskedLogSoftmax(logits, mask *Node, axes ...int) *Node {
...
+// be so small as to exceed the range of normal floating point numbers. In such cases the logarithm of the calculated probability is stored.
+// This function allows adding probabilities stored in such a fashion.
+func LogAddExp(x, y *Node) *Node {
+	xShape := x.Shape()
+	yShape := y.Shape()
+	xShape.Assert(yShape.DType, yShape.Dimensions...)
+
+	max := Max(x, y)
+	delta := Sub(x, y)
+	return Where(IsFinite(delta),
```
Comment (janpfeifer) [inline graph/ops_util.go]: This `Where(IsFinite(delta)...)` makes it stable on the forward path (in the direct calculation).

But it will still lead to NaNs on backward path, because the NaNs will appear in the "ifTrue" parameter of the `Where` clause (where `delta` is finite), and those NaNs are "contagious". I can explain to you in details if you want by VC.

But one way of solving this, is never allowing NaNs anywhere. Something like:

```
deltaFiniteMask := IsFinite(delta)
safeDelta := Where(deltaFiniteMask, delta, ScalarZero(g, dtype))
return Where(deltaFiniteMask,
		Add(max, Log1p(Exp(Neg(Abs(safeDelta))))),
		Add(x, y))
```

This way there will never be a `NaN` in any of the sides of the where. Where there is an infinite, there may be NaNs in other places of the graph, but this keeps this one safe :)

Btw: `g := x.Graph` and `dtype := x.DType()`

File: graph/ops_util.go [hunk 3]
```diff
@@ -190,6 +190,14 @@ func PositiveIndicator(x *Node) *Node {
 	return Sign(Add(Sign(x), one))
 }
 
+// NegativeIndicator returns 1 where x <= 0, 0 otherwise. See also StrictlyNegativeIndicatorIndicator.
+// E.g: NegativeIndicator({1.0, 0.0001, 0, -0.2, -3.0}) -> [0, 0, 1, 1, 1], with the same shape/dtype as x.
+func NegativeIndicator(x *Node) *Node {
```
Comment (janpfeifer) [inline graph/ops_util.go]: Quick note: this reminded me of a mistake I did:

`PositiveIndicator` and `NegativeIndicator` should be (using math jargon) when x > 0 and x < 0 respectively.

To include 0, the usual name would be `NonNegativeIndicator` (x >= 0) and `NonPositiveIndicator` (x <= 0).

I'm happy to fix the naming after I merge your code though (since this was my mistake).

File: ml/context/context.go [hunk 4]
```diff
@@ -392,10 +394,17 @@ func GetParamOr[T any](ctx *Context, key string, defaultValue T) T {
 		return value
 	}
 
-	// Try converting, for instance, a float32 could be converted to float64.
+	// Try T.TextUnmarshaler(valueAny)
 	v := reflect.ValueOf(valueAny)
 	typeOfT := reflect.TypeOf(defaultValue)
-	if !v.CanConvert(typeOfT) {
+	valueT := reflect.New(typeOfT)
```
Comment (janpfeifer) [inline ml/context/context.go:400]: Since you are here, would you mind adding one such conversion example in the `TestParams()` function, in `contest_test.go` file ?

**Review by janpfeifer — CHANGES_REQUESTED** [review]
Thank you Rener, really neat PR.

I made a couple of small requests, I hope they are ok ? 

And after them I merge and create a new release v0.17.1.

**Comment by janpfeifer** [comment]
Btw, just FYI, I have PR #122 coming up to fix L1  and L2 normlization in a couple of cases. But I'll wait to merge your code.

File: graph/ops_util.go (hunk 2 above)
Comment (janpfeifer) [reply graph/ops_util.go]: I just did a similar fix to `L2Normalize` in #122

File: graph/ops_util.go (hunk 3 above)
Comment (TuSKan) [reply graph/ops_util.go]: I just fixed the names and include tests for negatives
Where was _PositiveIndicator_ -> is now _NonNegativeIndicator_
Where was _StrictlyPositiveIndicator_ -> is now  _PositiveIndicator_
Where was NegativeIndicator_ -> is now  _NonPositiveIndicator_
Where was _StrictlyNegativeIndicator_ -> is now _NegativeIndicator_

**Comment by TuSKan** [comment]
Please, don't merge yet.

1. I wouldn't be able to fix classifier missing metrics error
2. triplet loss performance is absolutely awful for any strategy/distance variations

**Comment by janpfeifer** [comment]
Yes, so I haven't used triplet loss, but from colleagues I hear it's finicky to get it right -- but in the cases it works it great.

Also, I was looking at it yesterday, and at a later PR we could try to make its distances safe to 0s -- fully zero embeddings  will generate NaNs in the training (and infinity in inference) for at least the cosine similarity. That's what I fixed in the L2Normalization. 

Hey, would your rather have me merge your PR as is, and you fix the classifier in a next PR, or that I keep merging my (there is a new PR coming from another collaborator soon I think) changes to the GoMLX repo, and later you merge the latest changes ? 

cheers

**Comment by TuSKan** [comment]
Ok feel free to merge as you want, no problem

**Review by janpfeifer — APPROVED** [review]
Thanks for the renaming of the indicator functions!

**Comment by janpfeifer** [comment]
Hmm ... something went wrong ... your changes went to the `mnist` branch, as opposed to the `main` branch. Let me see if I can fix it.
